### PR TITLE
[Azure Search] Add new skill and cognitive services attachment to SDK

### PR DESCRIPTION
**Pull request by arv100kri**
## Description
* Regenerate the SDK to generate code for `EntityRecognitionSkill` and the ability to attach cognitive services to a skillset - this is from the REST spec merged via [this PR](https://github.com/Azure/azure-rest-api-specs/pull/5145)
* Mark `NamedEntityRecognitionSkill` as deprecated
* Add a `CosmosDb` data source creation method, to address #5060 and mark `DocumentDb` method as deprecated
* Expose the `jsonLines` indexer parsing mode which was recently introduced in the REST API
* Update tests for skills, skillsets and data sources

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

**Comment by arv100kri** [comment]
Does marking properties/methods/members constitute as a breaking change?
It shouldn't matter anyway, I think because we have bumped up the version number anyway, correct?

**Comment by brjohnstmsft** [comment]
@arv100kri Do you mean marking members with `[Obsolete]`? That isn't a breaking change in and of itself. And you're right; It doesn't matter anyway because we've bumped the major version number in anticipation of breaking changes for Complex Types.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Indexers/Models/IndexingParametersExtensions.cs [hunk 1]
```diff
@@ -93,6 +93,19 @@ public static IndexingParameters ExcludeFileNameExtensions(this IndexingParamete
         public static IndexingParameters ParseJson(this IndexingParameters parameters) =>
             Configure(parameters, ParsingModeKey, "json");
 
+        /// <summary>
+        /// Tells the indexer to assume that all blobs contain new-line sepatated JSON, which it will then parse such that individual JSON entities in each blob
```
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Indexers/Models/IndexingParametersExtensions.cs]: sepatated -> separated

**Review by bitsforthought — CHANGES_REQUESTED** [review]

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityRecognitionSkillLanguage.cs [hunk 2]
```diff
@@ -6,28 +6,14 @@ namespace Microsoft.Azure.Search.Models
...
     using Serialization;
+    using System;
 
     /// <summary>
     /// Defines the format of NamedEntityRecognitionSkill supported language codes.
     /// </summary>
     [JsonConverter(typeof(ExtensibleEnumConverter<NamedEntityRecognitionSkillLanguage>))]
     public sealed class NamedEntityRecognitionSkillLanguage : ExtensibleEnum<NamedEntityRecognitionSkillLanguage>
     {
-        /// <summary>
```
[the remark this answers is not in the document]
Comment (brjohnstmsft) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityRecognitionSkillLanguage.cs]: In fact, the entire type should be marked as `Obsolete`.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityCategory.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        [EnumMember(Value = "location")]
+        Location,
+        [EnumMember(Value = "organization")]
+        Organization,
+        [EnumMember(Value = "person")]
+        Person
+    }
+
+    [Obsolete]
```
[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityCategory.cs]: > @arv100kri Do you mean marking members with `[Obsolete]`? That isn't a breaking change in and of itself. And you're right; It doesn't matter anyway because we've bumped the major version number in anticipation of breaking changes for Complex Types.

Yup, and thanks for clarifying!

**Comment by arv100kri** [comment]
> @arv100kri Do you mean marking members with `[Obsolete]`? That isn't a breaking change in and of itself. And you're right; It doesn't matter anyway because we've bumped the major version number in anticipation of breaking changes for Complex Types.

Yup, and thanks for clarifying!

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityCategory.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityCategory.cs]: Yup, I should've added the [Obsolete] to the enum and the extension... will do

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityRecognitionSkillLanguage.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityRecognitionSkillLanguage.cs]: Will do (the entire type is already marked as Obsolete)
My only reasoning for removal was because those languages are not supported, but then again, if this entire type is Obsolete, this doesn't matter...

File: src/SDKs/Search/DataPlane/Search.Tests/Tests/SkillsetsTests.cs [hunk 4]
```diff
@@ -112,8 +112,19 @@ public void CreateSkillsetReturnsCorrectDefinitionOcrNamedEntity()
...
-                CreateAndValidateSkillset(searchClient, CreateTestSkillsetOcrNamedEntity(null, null));
-                CreateAndValidateSkillset(searchClient, CreateTestSkillsetOcrNamedEntity(TextExtractionAlgorithm.Printed, new List<NamedEntityCategory> { NamedEntityCategory.Location, NamedEntityCategory.Organization, NamedEntityCategory.Person } ));
+                CreateAndValidateSkillset(searchClient, CreateTestSkillsetOcrEntity(null, null));
+                CreateAndValidateSkillset(searchClient, CreateTestSkillsetOcrEntity(TextExtractionAlgorithm.Printed, new List<EntityCategory> { EntityCategory.Location, EntityCategory.Organization, EntityCategory.Person }));
+            });
+        }
+
+        // Purely to test the still existing, but obsolete NamedEntityRecognitionSkill
+        [Fact]
+        public void CreateSkillsetReturnsCorrectDefinitionNamedEntity()
```
[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Search.Tests/Tests/SkillsetsTests.cs]: > I've never seen tests marked with `Obsolete` before. Is this a way to suppress compiler warnings from using obsolete features?

Never mind :) I should just remove this and the other test as you suggest

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/DataSources/Models/DataSourceType.cs [hunk 5]
```diff
@@ -21,8 +22,14 @@ public sealed class DataSourceType : ExtensibleEnum<DataSourceType>
         /// <summary>
         /// Indicates a DocumentDB datasource.
         /// </summary>
+        [Obsolete("This member is deprecated. Use CosmosDb instead")]
         public static readonly DataSourceType DocumentDb = new DataSourceType("documentdb");
 
+        /// <summary>
+        /// Indicates a CosmosDB datasource.
+        /// </summary>
+        public static readonly DataSourceType CosmosDb = new DataSourceType("documentdb");
```
[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/DataSources/Models/DataSourceType.cs]: Unfortunately it is... all of Azure Search documentation refers to the data source as "CosmosDb", but in the indexer definition, the wire contract is "documentdb"

https://docs.microsoft.com/en-us/azure/search/search-howto-index-cosmosdb#step-1-create-a-data-source

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Generated/SkillsetsOperations.cs [hunk 6]
```diff
@@ -185,7 +185,7 @@ public async Task<AzureOperationResponse<Skillset>> CreateOrUpdateWithHttpMessag
 
             if (customHeaders != null)
             {
-                foreach(var _header in customHeaders)
+                foreach (var _header in customHeaders)
```
[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Generated/SkillsetsOperations.cs]: I think this is some VS weirdness as part of my commit `3bcaaf55`, where I replaced NERSkill with EntitySkill... that commit messed up the whitespacing in this file...

I'll regenerate this from swagger and add a different commit. Thanks for the spot!

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityRecognitionSkillLanguage.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (brjohnstmsft) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/Skillsets/Models/NamedEntityRecognitionSkillLanguage.cs]: Also, removing fields is a breaking change. The idea is to make them `[Obsolete]` for one release cycle before removing them outright.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Generated/SkillsetsOperations.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (brjohnstmsft) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Generated/SkillsetsOperations.cs]: Nobody should ever be editing the generated code by hand. Was this the result of a global find-and-replace or something?

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Common/Customizations/Search/Models/ExtensibleEnum.cs [hunk 7]
```diff
@@ -112,7 +112,8 @@ public override string ToString()
         private static Dictionary<string, T> CreateNameMap()
         {
             IEnumerable<FieldInfo> allPublicStaticFields =
-                typeof(T).GetTypeInfo().DeclaredFields.Where(f => f.IsStatic && f.IsPublic);
+                typeof(T).GetTypeInfo().DeclaredFields
+                .Where(f => f.IsStatic && f.IsPublic && f.GetCustomAttribute(typeof(ObsoleteAttribute), false) != null);
```
[the remark this answers is not in the document]
Comment (brjohnstmsft) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Common/Customizations/Search/Models/ExtensibleEnum.cs]: Discussed this offline. The issue is that now we have two properties with different names (`CosmosDb` and `DocumentDb`) but the same value on the wire (`"documentdb"`). This causes the call to `ToDictionary` to fail. Using the `[Obsolete]` attribute to disambiguate makes sense in this scenario. Let's go ahead with this change, with test coverage for this new case.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Generated/SkillsetsOperations.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Generated/SkillsetsOperations.cs]: Nope. AFAIK I must've just opened the file in VS to view it and then must've hit CTRL-S and hit quit... Some plugin that i have that does this auto whitespace formatting would've recked havoc...

What is expected however is the "re-ordering" of the methods, because I changed the order in swagger to be consistent with other operations

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/DataSources/Models/DataSourceType.cs [hunk 8]
```diff
@@ -21,8 +22,14 @@ public sealed class DataSourceType : ExtensibleEnum<DataSourceType>
         /// <summary>
         /// Indicates a DocumentDB datasource.
         /// </summary>
+        [Obsolete("This member is deprecated. Use CosmosDb instead")]
```
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Service/Customizations/DataSources/Models/DataSourceType.cs]: "This member is deprecated" is redundant.

**Review by bitsforthought — CHANGES_REQUESTED** [review]
Almost there, just a few more minor changes

File: src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs [hunk 9]
```diff
@@ -69,6 +72,23 @@ public void CannotCompareDifferentEnumsForEquality()
...
 
+        [Fact]
+        public void CanCreateNonObsoleteTypeFromValue()
+        {
+            TestEnumWithObsolete notObsolete = TestEnumWithObsolete.Create("NotObsolete");
+            Assert.Same(TestEnumWithObsolete.NotObsolete, notObsolete);
+        }
+
+        [Fact]
+        public void CannotCreateObsoleteTypeFromValue()
```
[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs]: The test above was for that... you don’t think that accurately tests what I want?

[the remark this answers is not in the document]
Comment (brjohnstmsft) [reply src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs]: @arv100kri Sorry, I missed that before. It's a bit hard to parse since "NotObsolete" maps to the enum member `Obsolete`. Maybe some renaming would make this more obvious.

[the remark this answers is not in the document]
Comment (arv100kri) [reply src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs]: Naming is hard... I've updated the values and certain property names, let me know what you think?

File: src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs [hunk 10]
```diff
@@ -113,10 +113,10 @@ private class TestEnum2 : ExtensibleEnum<TestEnum2>
...
         {
-            public static readonly TestEnumWithObsolete NotObsolete = new TestEnumWithObsolete("NotObsolete");
+            public static readonly TestEnumWithObsolete RecommendedPropertyValue = new TestEnumWithObsolete("Value");
 
             [Obsolete]
-            public static readonly TestEnumWithObsolete Obsolete = new TestEnumWithObsolete("NotObsolete");
+            public static readonly TestEnumWithObsolete DeprecatedPropertyValue = new TestEnumWithObsolete("Value");
 
             [Obsolete]
             public static readonly TestEnumWithObsolete Obsolete2 = new TestEnumWithObsolete("Obsolete");
```
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs]: What is the presence of `Obsolete2` testing?

File: src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs [hunk 11]
```diff
@@ -85,7 +85,7 @@ public void CannotCreateObsoleteTypeFromValue()
             // this should result in a new TestEnumWithObsolete instance being created
             TestEnumWithObsolete obsolete = TestEnumWithObsolete.Create("Obsolete");
```
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Search.Tests/Tests/ExtensibleEnumTests.cs]: This test no longer makes sense... Why would it ever be `DeprecatedPropertyValue` when they have different values...? At least before, the value "Obsolete" matched the name `Obsolete`... Now I'm confused as to what's being tested here.

**Review by bitsforthought — CHANGES_REQUESTED** [review]
Still confused by the tests :(

**Review by bitsforthought — APPROVED** [review]
